tests/unit/wait/waitDevice.cpp

```diff
@@ -0,0 +1,127 @@
+#include <alpaka/onHost/executeForEach.hpp>
+
+#include <catch2/catch_template_test_macros.hpp>
+#include <catch2/catch_test_macros.hpp>
+
+using namespace alpaka;
+
+// any value for testing
+int const testValue = 49;
+
+using TestApis = std::decay_t<decltype(onHost::allBackends(onHost::enabledApis, onHost::example::enabledExecutors))>;
+
+// Simple write kernel used to create actual device work for testing wait(device).
+struct WaitTestWriteKernel
+{
+    template<typename TAcc, typename TBuf>
+    ALPAKA_FN_ACC void operator()(TAcc const&, TBuf out) const
+    {
+        out[0] = testValue;
+    }
+};
+
+template<typename T_DataType>
+void executeWaitTest(auto& device, auto const& exec)
+{
+    using DataType = T_DataType;
+
+    // Test 1: Multiple concurrent queues
+    constexpr uint32_t numQueues = 3;
+    std::vector<decltype(device.makeQueue())> queues;
+    std::vector<decltype(onHost::alloc<DataType>(device, Vec<uint32_t, 1>{1u}))> devBufs;
+    std::vector<decltype(onHost::allocHostLike(onHost::alloc<DataType>(device, Vec<uint32_t, 1>{1u})))> hostBufs;
+
+    // allocate memory and create queues
+    constexpr auto extent = Vec<uint32_t, 1>{1u};
+    for(uint32_t i = 0; i < numQueues; ++i)
+    {
+        queues.push_back(device.makeQueue());
+        devBufs.push_back(onHost::alloc<DataType>(device, extent));
+        hostBufs.push_back(onHost::allocHostLike(devBufs.back()));
+    }
+
+    // enqueue work on multiple queues concurrently
+    constexpr auto frameSize = Vec<uint32_t, 1>{1u};
+    for(uint32_t i = 0; i < numQueues; ++i)
+    {
+        auto kernelBundle = KernelBundle{WaitTestWriteKernel{}, devBufs[i]};
+        queues[i].enqueue(exec, onHost::FrameSpec{extent, frameSize}, kernelBundle);
+        // Note: We deliberately don't wait on individual queues here
+    }
+
+    // Test 2: Multiple async operations on SAME queue
+    auto singleQueue = device.makeQueue();
+    constexpr uint32_t numOpsPerQueue = 3;
+    std::vector<decltype(onHost::alloc<DataType>(device, extent))> singleQueueBufs;
+    std::vector<decltype(onHost::allocHostLike(onHost::alloc<DataType>(device, extent)))> singleQueueHostBufs;
+
+    for(uint32_t i = 0; i < numOpsPerQueue; ++i)
+    {
+        singleQueueBufs.push_back(onHost::alloc<DataType>(device, extent));
+        singleQueueHostBufs.push_back(onHost::allocHostLike(singleQueueBufs.back()));
+
+        // Enqueue multiple operations on same queue - they queue up asynchronously
+        auto kernelBundle = KernelBundle{WaitTestWriteKernel{}, singleQueueBufs[i]};
+        singleQueue.enqueue(exec, onHost::FrameSpec{extent, frameSize}, kernelBundle);
+    }
+
+    // TEST: wait for device to complete ALL work on ALL queues (including single queue ops)
+    REQUIRE_NOTHROW(onHost::wait(device));
+
+    // verify all kernels executed correctly on all queues
+    for(uint32_t i = 0; i < numQueues; ++i)
+    {
+        onHost::memcpy(queues[i], hostBufs[i], devBufs[i]);
```

Comment:
this memcpy should be in the for loop where you push the work.
The reason is that `onHost::wait(queues[i]);` is syncronizing each queue and even if you skip wait(device) all will work.